libraries/lib-wave-track/WaveClip.cpp

```diff
@@ -198,6 +198,55 @@ void WaveClip::SetSamples(size_t ii,
    MarkChanged();
 }
 
+bool WaveClip::GetFloatAtTime(
+   double t, size_t iChannel, float& value, bool mayThrow) const
+{
+   if (!WithinPlayRegion(t))
+      return false;
+   const auto start = TimeToSamples(t);
+   return GetSamples(
+      iChannel, reinterpret_cast<samplePtr>(&value), floatSample, start, 1u,
+      mayThrow);
+}
+
+void WaveClip::SetFloatsFromTime(
+   double t, size_t iChannel, const float* buffer, size_t numFloats,
+   sampleFormat effectiveFormat)
+{
+   const auto maybeNegativeStart = TimeToSamples(t);
+   const auto maybeOutOfBoundEnd = maybeNegativeStart + numFloats;
+   const auto effectiveStart = std::max(sampleCount { 0 }, maybeNegativeStart);
+   const auto effectiveEnd =
+      std::min(GetVisibleSampleCount(), maybeOutOfBoundEnd);
+   // Cannot be greater than `numFloats` -> safe cast
+   const auto effectiveLen = (effectiveEnd - effectiveStart).as_size_t();
+   if (effectiveLen == 0)
+      return;
+   // Cannot be greater than `numFloats` -> safe cast
+   const auto numLeadingZeros =
+      (effectiveStart - maybeNegativeStart).as_size_t();
```

Comment:
Actually, your informal proof here is wrong.  What if `t` happens to be a very negative value?

So to be safe you should write
`const auto numLeadingZeros = limitSampleBufferSize(numFloats, effectiveStart - maybeNegativeStart);`

Reply:
By substitution, `effectiveStart - maybeNegativeStart == std::max(sampleCount { 0 }, maybeNegativeStart) - maybeNegativeStart`,
and `maybeNegativeStart = TimeToSamples(t)`. So if `t < 0`, then `effectiveStart - maybeNegativeStart = - maybeNegativeStart`, that is, a positive value.

For `t` to be negative is actually not a corner case, here, and the code has proven to work well as far as my testing went, which is another evidence that it is correct.